server/chat-plugins/github.ts

```diff
@@ -0,0 +1,266 @@
+const COOLDOWN = 10 * 60 * 1000;
+
+export const gitData: GitData = JSON.parse(FS("config/chat-plugins/github.json").readIfExistsSync() || "{}");
+
+interface GitHookHandler {
+	on(event: 'pull_request', callback: (repo: string, ref: string | undefined, result: PullRequest) => void): void;
+	on(event: 'push', callback: (repo: string, ref: string, result: Push) => void): void;
+	on(event: string, callback: (repo: string, ref: string | undefined, result: any) => void): void;
+	listen(): void;
+	server: import('http').Server;
+}
+
+interface Push {
+	commits: Commit[];
+	sender: {login: string};
+	compare: string;
+}
+
+interface PullRequest {
+	action: string;
+	number: number;
+	pull_request: {
+		url: string,
+		html_url: string,
+		title: string,
+		user: {
+			login: string,
+			html_url: string,
+		},
+		merge_commit_sha: string,
+	};
+	sender: {login: string};
+}
+
+interface Commit {
+	id: string;
+	message: string;
+	author: {name: string, avatar_url: string};
+	url: string;
+}
+
+interface GitData {
+	usernames?: {[username: string]: string};
+	bans?: {[username: string]: string};
+}
+
+export const GitHub = new class {
+	readonly hook: GitHookHandler | null = null;
+	updates: {[k: string]: number} = Object.create(null);
+	constructor() {
+		try {
+			this.hook = (require('githubhook'))(Config.github);
+		} catch {}
+		this.listen();
+	}
+	listen() {
+		if (!this.hook) return;
+		this.hook.listen();
+		this.hook.on('push', (repo, ref, result) => this.handlePush(repo, ref, result));
+		this.hook.on('pull_request', (repo, ref, result) => this.handlePull(repo, ref, result));
+	}
+	private getRepoName(repo: string) {
+		switch (repo) {
+		case 'pokemon-showdown':
+			return 'server';
+		case 'pokemon-showdown-client':
+			return 'client';
+		case 'Pokemon-Showdown-Dex':
+			return 'dex';
+		default:
+			return repo.toLowerCase();
+		}
+	}
+	handlePush(repo: string, ref: string, result: Push) {
+		const branch = /[^/]+$/.exec(ref)?.[0] || "";
+		if (branch !== 'master') return;
+		const messages = {
+			staff: [] as string[],
+			development: [] as string[],
+		};
+		for (const commit of result.commits) {
+			const {message, url} = commit;
+			let shortened = /.+/.exec(message)?.[0] || "";
+			if (message !== shortened) {
+				shortened += '…';
+			}
+			const username = this.getUsername(commit.author.name);
+			const repoName = this.getRepoName(repo);
+			const id = commit.id.substring(0, 6);
+			messages.development.push(
+				Utils.html`[<span style="color:#FF00FF">${repoName}</span>] <a href=\"${url}\"><span style="color:#606060">${id}</span></a> ${shortened} <span style="color:#909090">(${username})</span>`
+			);
+			messages.staff.push(Utils.html`[<span style="color:#FF00FF">${repoName}</span>] <a href=\"${url}\">${shortened}</a> <span style="color:#909090">(${username})</span>`);
+		}
+		for (const k in messages) {
+			this.report(k as 'staff', repo, messages[k as 'staff']);
```

Comment:
Casting to `'staff'` is a little odd; why not just hardcode the `staff` and `development` rooms, or make the `report` method accept a roomid:Array<string> object?

Reply:
I chose to cast to Staff since it was that or just casting to 'staff' | 'development'. I didn't really see the need to cast to both and make it a little bit more annoying to read.

Reply:
My suggestions remove the need for casting altogether.

Reply:
>  roomid:Array object?

Would still require casting, no?

> why not just hardcode the staff and development rooms

I don't understand how this is different?

Reply:
> Would still require casting, no?

It shouldn't?
```ts
report(messages: {[k: string]: string[] | string}, repo: string) {
	for (const [roomid, message] of Object.entries(messages)) {
		if (!STAFF_REPOS.includes(repo) && roomid === 'staff') return;
		if (Array.isArray(message)) message = message.join('<br />');
		Rooms.get(roomid)?.add(`|html|<div class="infobox">${message}</div>`).update();
	}
}
```

Reply:
I just made it cast to `RoomID`.